src/dupin.cpp

```diff
@@ -42,51 +42,58 @@ void DynamicProgramming::regression_setup(linear_fit_struct &lfit) {
   lfit.y = data;
 }
 
-Eigen::VectorXd DynamicProgramming::regression_line(int start, int end, int dim,
-                                             linear_fit_struct &lfit) {
-  int n = end - start;
-  Eigen::VectorXd x = lfit.x.segment(start, n);
-  Eigen::VectorXd y = lfit.y.col(dim).segment(start, n);
+//work in progress, the rowwise colwise is messing up
+Eigen::MatrixXd DynamicProgramming::regression_lines(int start, int end, linear_fit_struct &lfit) {
+    int n = end - start;
+    Eigen::VectorXd x = lfit.x.segment(start, n);
+    Eigen::MatrixXd y = lfit.y.block(start, 0, n, num_features);
+
+    // Ensure x is in a two-dimensional form for broadcasting
+    Eigen::MatrixXd x_matrix = x.replicate(1, num_features);
+
+    // Calculate means
+    double x_mean = x.mean();
+    Eigen::VectorXd y_mean = y.colwise().mean();
+
+    // Center the data around 0
+    Eigen::MatrixXd x_centered = x_matrix.colwise() - Eigen::VectorXd::Constant(n, x_mean);
+    Eigen::MatrixXd y_centered = y.rowwise() - y_mean.transpose();
 
-  double x_mean = x.mean();
-  double y_mean = y.mean();
+    // Calculate slopes for each feature
+    Eigen::VectorXd slope = (x_centered.array() * y_centered.array()).colwise().sum() / x_centered.array().square().sum();
 
-  Eigen::VectorXd x_centered = x.array() - x_mean;
-  Eigen::VectorXd y_centered = y.array() - y_mean;
+    // Calculate intercepts for each feature
+    Eigen::VectorXd intercept = y_mean.array() - slope.array() * x_mean;
 
-  double slope = x_centered.dot(y_centered) / x_centered.squaredNorm();
-  double intercept = y_mean - slope * x_mean;
+    // everything till this line is functioning fine; I might be overcomplicating it
+    Eigen::MatrixXd regression_lines = (x_matrix.array().colwise() - x_mean).colwise() * slope.array() + intercept.transpose().array();
```

Comment:
I think you just need to do something like `x_matrix * slope.replicate(x.size(), 1) + interept.replicate(x.size(), 1)` though I am not too familiar with Eigen.